booster/library/Booster/Pattern/Rewrite.hs

```diff
@@ -801,108 +792,113 @@ performRewrite doTracing def mLlvmLibrary smtSolver varsToAvoid mbMaxDepth cutLa
-                                else withSimplified pat' "Retrying with simplified pattern" (doSteps True)
-                        Right (trivial@RewriteTrivial{}, _) -> withPatternContext pat' $ do
-                            logMessage $ "Simplified to bottom after " <> showCounter counter
-                            pure trivial
-                        Right (aborted@RewriteAborted{}, _) -> withPatternContext pat' $ do
-                            logMessage $ "Aborted after " <> showCounter counter
-                            simplifyResult pat' aborted
-                        -- if unification was unclear and the pattern was
-                        -- unsimplified, simplify and retry rewriting once
-                        Left failure@(RuleApplicationUnclear rule _ remainder)
-                            | not wasSimplified -> do
-                                emitRewriteTrace $ RewriteStepFailed failure
-                                -- simplify remainders, substitute and rerun.
-                                -- If failed, do the pattern-wide simplfication and rerun again
-                                withSimplified pat' "Retrying with simplified pattern" (doSteps True)
-                            | otherwise -> do
-                                -- was already simplified, emit an abort log entry
-                                withRuleContext rule . withContext CtxMatch . withContext CtxAbort $
-                                    getPrettyModifiers >>= \case
-                                        ModifiersRep (_ :: FromModifiersT mods => Proxy mods) ->
-                                            logMessage $
-                                                WithJsonMessage (object ["remainder" .= (bimap externaliseTerm externaliseTerm <$> remainder)]) $
-                                                    renderOneLineText $
-                                                        "Uncertain about match with rule. Remainder:"
-                                                            <+> ( hsep $
-                                                                    punctuate comma $
-                                                                        map (\(t1, t2) -> pretty' @mods t1 <+> "==" <+> pretty' @mods t2) $
-                                                                            NE.toList remainder
-                                                                )
-                                emitRewriteTrace $ RewriteStepFailed failure
+
+        case counter of
+            c
+                | depthReached c -> do
+                    logDepth $ "Reached maximum depth of " <> maybe "?" showCounter mbMaxDepth
+                    (if wasSimplified then pure else simplifyResult pat') $ RewriteFinished Nothing Nothing pat'
+                | counter > 0
+                , not wasSimplified
+                , maybe False ((== 0) . (counter `mod`)) mbSimplify -> do
```

Comment:
Please consider factoring this out into a function, I think it's a little too much going on here.